File: nim/src/dependency/constructor.nim
# internal
import fibonacci
import square


type
    Constructor* = ref object
        fibonacci: Fibonacci
        square: Square

proc newConstructor*(fibonacci: Fibonacci, square: Square): Constructor =
    return Constructor(
        fibonacci: fibonacci,
        square: square
    )

proc doAll*(self: Constructor, num: uint32) =
    if (self.fibonacci == nil or
            self.square == nil):
        return

    let fibonacciResult = self.fibonacci.compute(num)
    echo "Fibonacci of " & $num & ": " & $fibonacciResult

    let squareResult = self.square.compute(num)
    echo "Square of " & $num & ": " & $squareResult

File: python/src/dependency/constructor.py
# internal
from dependency.fibonacci import Fibonacci
from dependency.square import Square


class Constructor(object):
    __fibonacci: Fibonacci = None
    __square: Square = None

    def __init__(self, fibonacci: Fibonacci, square: Square):
        self.__fibonacci = fibonacci
        self.__square = square

    def do_all(self, num: int):
        if (self.__fibonacci == None
                or self.__square == None):
            return
        
        fibonacci_result: int = self.__fibonacci.compute(num)
        print("Fibonacci of " + str(num) + ": " + str(fibonacci_result))
        
        square_result: int = self.__square.compute(num)
        print("Square of " + str(num) + ": " + str(square_result))

    def do_fibonacci(self, num: int):
        if self.__fibonacci == None:
            return
        
        fibonacci_result: int = self.__fibonacci.compute(num)
        print("Fibonacci of " + str(num) + ": " + str(fibonacci_result))

    def do_square(self, num: int):
        if self.__square == None:
            return
        
        square_result: int = self.__square.compute(num)
        print("Square of " + str(num) + ": " + str(square_result))